File: vignettes/06_middleware/python/06_middleware.py
"""
Middleware — Intercepting and modifying agent behaviour (Python)

This sample demonstrates how to use middleware to add logging, timing,
and guardrail capabilities to agents at three levels: agent, chat client,
and function invocation. It mirrors the Julia vignette 06_middleware.

Prerequisites:
  - Ollama running locally with `qwen3:8b` pulled
  - pip install agent-framework-ollama
"""

import asyncio
import time
from typing import Awaitable, Callable

from agent_framework import (
    Agent,
    AgentContext,
    AgentMiddleware,
    AgentResponse,
    ChatContext,
    ChatMiddleware,
    FunctionInvocationContext,
    FunctionMiddleware,
    Message,
    tool,
)
from agent_framework.ollama import OllamaChatClient


# ── Define a tool ────────────────────────────────────────────────────────

@tool(approval_mode="never_require")
def get_population(city: str) -> str:
    """Get the population of a city.

    Args:
        city: The city name.

    Returns:
        The population as a string.
    """
    populations = {
        "Paris": "2.1 million",
        "London": "8.8 million",
        "Tokyo": "14 million",
    }
    return populations.get(city, "Unknown")


# ── Agent middleware: logging ────────────────────────────────────────────

class LoggingAgentMiddleware(AgentMiddleware):
    """Logs the start and end of every agent run."""

    async def process(
        self,
        context: AgentContext,
        call_next: Callable[[], Awaitable[None]],
    ) -> None:
        agent_name = context.agent.name
        n_messages = len(context.messages)
        print(f"[AgentMW] Starting {agent_name} with {n_messages} message(s)")

        await call_next()

        print(f"[AgentMW] {agent_name} completed")


# ── Chat middleware: timing ──────────────────────────────────────────────

class TimingChatMiddleware(ChatMiddleware):
    """Measures the duration of each LLM API call."""

    async def process(
        self,
        context: ChatContext,
        call_next: Callable[[], Awaitable[None]],
    ) -> None:
        n_msgs = len(context.messages)
        print(f"  [ChatMW] Sending {n_msgs} messages to LLM...")

        start = time.time()
        await call_next()
        elapsed = time.time() - start

        print(f"  [ChatMW] LLM responded in {elapsed:.2f}s")


# ── Function middleware: tool logging ────────────────────────────────────

class ToolLoggingMiddleware(FunctionMiddleware):
    """Logs every tool invocation with its arguments and result."""

    async def process(
        self,
        context: FunctionInvocationContext,
        call_next: Callable[[], Awaitable[None]],
    ) -> None:
        tool_name = context.function.name
        print(f"    [FuncMW] Calling tool: {tool_name}")

        await call_next()

        print(f"    [FuncMW] Tool {tool_name} returned: {context.result}")


# ── Security guardrail middleware ────────────────────────────────────────

class SecurityMiddleware(AgentMiddleware):
    """Blocks requests containing sensitive keywords."""

    async def process(
        self,
        context: AgentContext,
        call_next: Callable[[], Awaitable[None]],
    ) -> None:
        for msg in context.messages:
            if msg.text and any(
                kw in msg.text.lower() for kw in ("password", "secret")
            ):
                print("[Security] Blocked request: sensitive content detected")
                context.result = AgentResponse(
                    messages=[
                        Message(
                            "assistant",
                            ["I cannot process requests containing sensitive information."],
                        )
                    ]
                )
                return  # short-circuit — do not call next

        await call_next()


async def main() -> None:
    client = OllamaChatClient(
        host="http://localhost:11434",
        model_id="qwen3:8b",
    )

    # ── Agent with all three middleware layers ────────────────────────────
    print("=== Full middleware pipeline ===")
    agent = client.as_agent(
        name="CityBot",
        instructions="You are a helpful city information assistant. Use the get_population tool when asked about population.",
        tools=[get_population],
        middleware=[
            LoggingAgentMiddleware(),
            TimingChatMiddleware(),
            ToolLoggingMiddleware(),
        ],
    )

    result = await agent.run("What is the population of Paris?")
    print(f"\nAnswer: {result.text}")
    print()

    # ── Security guardrail ───────────────────────────────────────────────
    print("=== Security guardrail ===")
    secure_agent = client.as_agent(
        name="SecureBot",
        instructions="You are a helpful assistant.",
        middleware=[SecurityMiddleware()],
    )

    result = await secure_agent.run("What is my password?")
    print(f"Blocked: {result.text}")

    result = await secure_agent.run("What is the capital of France?")
    print(f"Allowed: {result.text}")


if __name__ == "__main__":
    asyncio.run(main())

File: vignettes/06_middleware/csharp/06_middleware.cs
// Middleware — Intercepting and modifying agent behaviour (C#)
//
// This sample demonstrates how to use middleware to add logging, timing,
// and guardrail capabilities to agents. It mirrors the Julia vignette
// 06_middleware.
//
// Prerequisites:
//   - Ollama running locally with qwen3:8b pulled
//   - dotnet restore

using System.ComponentModel;
using System.Diagnostics;
using Microsoft.Agents.AI;
using Microsoft.Extensions.AI;
using OllamaSharp;

var endpoint = Environment.GetEnvironmentVariable("OLLAMA_ENDPOINT")
    ?? "http://localhost:11434";
var modelName = Environment.GetEnvironmentVariable("OLLAMA_MODEL_NAME")
    ?? "qwen3:8b";

// ── Define a tool ────────────────────────────────────────────────────────

[Description("Get the population of a city.")]
static string GetPopulation(
    [Description("The city name.")] string city) =>
    city switch
    {
        "Paris" => "2.1 million",
        "London" => "8.8 million",
        "Tokyo" => "14 million",
        _ => "Unknown"
    };

// ── Create base agent with tool ──────────────────────────────────────────

AIAgent baseAgent = new OllamaApiClient(new Uri(endpoint), modelName)
    .AsAIAgent(
        instructions: "You are a helpful city information assistant. Use the GetPopulation tool when asked about population.",
        name: "CityBot",
        tools: [AIFunctionFactory.Create(GetPopulation)]);

// ── Function invocation middleware: tool logging ─────────────────────────

async ValueTask<object?> ToolLoggingMiddleware(
    AIAgent agent,
    FunctionInvocationContext context,
    Func<FunctionInvocationContext, CancellationToken, ValueTask<object?>> next,
    CancellationToken cancellationToken)
{
    Console.WriteLine($"    [FuncMW] Calling tool: {context.Function.Name}");

    var result = await next(context, cancellationToken);

    Console.WriteLine($"    [FuncMW] Tool {context.Function.Name} returned: {result}");
    return result;
}

// ── Agent middleware: logging ────────────────────────────────────────────

async Task<AgentResponse> LoggingAgentMiddleware(
    IEnumerable<ChatMessage> messages,
    AgentSession? session,
    AgentRunOptions? options,
    AIAgent innerAgent,
    CancellationToken cancellationToken)
{
    var msgCount = messages.Count();
    Console.WriteLine($"[AgentMW] Starting {innerAgent.Name} with {msgCount} message(s)");

    var response = await innerAgent.RunAsync(messages, session, options, cancellationToken);

    Console.WriteLine($"[AgentMW] {innerAgent.Name} completed");
    return response;
}

// ── Chat client middleware: timing ───────────────────────────────────────

async Task<ChatResponse> TimingChatMiddleware(
    IEnumerable<ChatMessage> messages,
    ChatOptions? options,
    IChatClient innerClient,
    CancellationToken cancellationToken)
{
    var msgCount = messages.Count();
    Console.WriteLine($"  [ChatMW] Sending {msgCount} messages to LLM...");

    var sw = Stopwatch.StartNew();
    var response = await innerClient.GetResponseAsync(messages, options, cancellationToken);
    sw.Stop();

    Console.WriteLine($"  [ChatMW] LLM responded in {sw.Elapsed.TotalSeconds:F2}s");
    return response;
}

// ── Security guardrail middleware ────────────────────────────────────────

async Task<AgentResponse> SecurityMiddleware(
    IEnumerable<ChatMessage> messages,
    AgentSession? session,
    AgentRunOptions? options,
    AIAgent innerAgent,
    CancellationToken cancellationToken)
{
    foreach (var msg in messages)
    {
        var text = msg.Text?.ToLowerInvariant() ?? "";
        if (text.Contains("password") || text.Contains("secret"))
        {
            Console.WriteLine("[Security] Blocked request: sensitive content detected");
            return new AgentResponse([new ChatMessage(ChatRole.Assistant,
                "I cannot process requests containing sensitive information.")]);
        }
    }

    return await innerAgent.RunAsync(messages, session, options, cancellationToken);
}

// ── Build agent with full middleware pipeline ────────────────────────────

Console.WriteLine("=== Full middleware pipeline ===");

var middlewareAgent = baseAgent
    .AsBuilder()
    .Use(ToolLoggingMiddleware)
    .Use(LoggingAgentMiddleware, null)
    .Build();

Console.WriteLine(await middlewareAgent.RunAsync("What is the population of Paris?"));
Console.WriteLine();

// ── Security guardrail demo ──────────────────────────────────────────────

Console.WriteLine("=== Security guardrail ===");

AIAgent secureAgent = new OllamaApiClient(new Uri(endpoint), modelName)
    .AsAIAgent(
        instructions: "You are a helpful assistant.",
        name: "SecureBot");

var guardedAgent = secureAgent
    .AsBuilder()
    .Use(SecurityMiddleware, null)
    .Build();

// This request is blocked.
Console.WriteLine($"Blocked: {await guardedAgent.RunAsync("What is my password?")}");

// This request goes through.
Console.WriteLine($"Allowed: {await guardedAgent.RunAsync("What is the capital of France?")}");
